File: backend/app/core/enums.py
"""Domain enumerations.

Every enum is a ``StrEnum`` stored in a ``text`` column guarded by a CHECK
constraint built from :meth:`DomainEnum.values`. Native PostgreSQL ``ENUM``
types were rejected deliberately: adding a value to one cannot be done in the
same transaction that uses it, and removing one requires recreating the type
and every dependent column. A CHECK constraint is altered by a single, fully
reversible migration statement — which matters for a product that will keep
adding opportunity types, submission methods and publisher categories.
"""

from __future__ import annotations

from enum import StrEnum


class DomainEnum(StrEnum):
    """Base class adding the helpers migrations and schemas need."""

    @classmethod
    def values(cls) -> tuple[str, ...]:
        return tuple(member.value for member in cls)

    @classmethod
    def check_constraint(cls, column: str) -> str:
        """Render a SQL ``IN`` predicate for a CHECK constraint."""
        rendered = ", ".join(f"'{value}'" for value in cls.values())
        return f"{column} IN ({rendered})"


# --------------------------------------------------------------------------- #
# Identity and access
# --------------------------------------------------------------------------- #


class TenantStatus(DomainEnum):
    ACTIVE = "ACTIVE"
    SUSPENDED = "SUSPENDED"
    ARCHIVED = "ARCHIVED"


class MembershipStatus(DomainEnum):
    """Only ``ACTIVE`` grants access; the others are membership records that
    deliberately do not."""

    ACTIVE = "ACTIVE"
    INVITED = "INVITED"
    SUSPENDED = "SUSPENDED"
    REMOVED = "REMOVED"


class SystemRoleSlug(DomainEnum):
    """Slugs of the roles seeded into every new tenant.

    Authorisation never branches on these — they exist so seeding is
    deterministic and so the UI can label a role. Permission checks always go
    through the role→permission mapping.
    """

    OWNER = "owner"
    ADMIN = "admin"
    SEO_MANAGER = "seo_manager"
    SEO_SPECIALIST = "seo_specialist"
    VIEWER = "viewer"


# --------------------------------------------------------------------------- #
# Client websites and campaigns
# --------------------------------------------------------------------------- #


class ClientWebsiteStatus(DomainEnum):
    ACTIVE = "ACTIVE"
    PAUSED = "PAUSED"
    ARCHIVED = "ARCHIVED"


class CampaignStatus(DomainEnum):
    DRAFT = "DRAFT"
    ACTIVE = "ACTIVE"
    PAUSED = "PAUSED"
    COMPLETED = "COMPLETED"
    ARCHIVED = "ARCHIVED"


# --------------------------------------------------------------------------- #
# Publishers
# --------------------------------------------------------------------------- #


class PricingType(DomainEnum):
    """Whether a publisher charges for a listing.

    MVP business rule: only ``FREE`` may enter the submission workflow. The
    other values exist so a paid or unknown publisher can be *recorded* during
    research without becoming submittable — enforced in the service layer, by
    a database trigger on the submission path, and by
    :meth:`submittable_values`.
    """

    FREE = "FREE"
    PAID = "PAID"
    MIXED = "MIXED"
    UNKNOWN = "UNKNOWN"

    @classmethod
    def submittable_values(cls) -> tuple[str, ...]:
        return (cls.FREE.value,)


class PublisherStatus(DomainEnum):
    DISCOVERED = "DISCOVERED"
    QUALIFYING = "QUALIFYING"
    QUALIFIED = "QUALIFIED"
    REJECTED = "REJECTED"
    BLOCKED = "BLOCKED"
    ARCHIVED = "ARCHIVED"


class SubmissionMethod(DomainEnum):
    """How a listing reaches a publisher.

    ``FORM`` is opt-in per publisher and the form adapter refuses any target
    that presents a CAPTCHA or anti-bot challenge, falling back to ``MANUAL``.
    No value here implies bypassing a publisher's controls.
    """

    MANUAL = "MANUAL"
    FORM = "FORM"
    EMAIL = "EMAIL"
    API = "API"
    ACCOUNT_REQUIRED = "ACCOUNT_REQUIRED"
    UNKNOWN = "UNKNOWN"


class LinkType(DomainEnum):
    DOFOLLOW = "DOFOLLOW"
    NOFOLLOW = "NOFOLLOW"
    MIXED = "MIXED"
    UNKNOWN = "UNKNOWN"


class PublisherCategory(DomainEnum):
    """Legitimate free-listing categories supported by the MVP.

    The list contains only directory and profile listing types. Paid guest
    posts, sponsored articles, link marketplaces, PBNs and comment/forum spam
    are out of scope by design and have no representation here.
    """

    BUSINESS_DIRECTORY = "BUSINESS_DIRECTORY"
    LOCAL_DIRECTORY = "LOCAL_DIRECTORY"
    COMPANY_LISTING = "COMPANY_LISTING"
    STARTUP_DIRECTORY = "STARTUP_DIRECTORY"
    SOFTWARE_DIRECTORY = "SOFTWARE_DIRECTORY"
    INDUSTRY_DIRECTORY = "INDUSTRY_DIRECTORY"
    ORGANIZATION_LISTING = "ORGANIZATION_LISTING"
    PROFILE_LISTING = "PROFILE_LISTING"
    REVIEW_PLATFORM = "REVIEW_PLATFORM"
    EVENT_LISTING = "EVENT_LISTING"
    JOB_BOARD = "JOB_BOARD"
    OTHER = "OTHER"


class DiscoveryRunStatus(DomainEnum):
    PENDING = "PENDING"
    RUNNING = "RUNNING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


# --------------------------------------------------------------------------- #
# Opportunities
# --------------------------------------------------------------------------- #


class OpportunityType(DomainEnum):
    FREE_DIRECTORY_LISTING = "FREE_DIRECTORY_LISTING"
    FREE_LOCAL_LISTING = "FREE_LOCAL_LISTING"
    FREE_COMPANY_PROFILE = "FREE_COMPANY_PROFILE"
    FREE_STARTUP_LISTING = "FREE_STARTUP_LISTING"
    FREE_SOFTWARE_LISTING = "FREE_SOFTWARE_LISTING"
    FREE_INDUSTRY_LISTING = "FREE_INDUSTRY_LISTING"
    FREE_ORGANIZATION_LISTING = "FREE_ORGANIZATION_LISTING"
    FREE_PROFILE_LISTING = "FREE_PROFILE_LISTING"


class OpportunityStatus(DomainEnum):
    DISCOVERED = "DISCOVERED"
    QUALIFYING = "QUALIFYING"
    QUALIFIED = "QUALIFIED"
    REJECTED = "REJECTED"
    SELECTED = "SELECTED"
    READY = "READY"
    SUBMITTED = "SUBMITTED"
    PUBLISHED = "PUBLISHED"
    FAILED = "FAILED"
    EXPIRED = "EXPIRED"


# --------------------------------------------------------------------------- #
# Submissions
# --------------------------------------------------------------------------- #


class SubmissionStatus(DomainEnum):
    READY = "READY"
    IN_PROGRESS = "IN_PROGRESS"
    SUBMITTED = "SUBMITTED"
    PENDING_APPROVAL = "PENDING_APPROVAL"
    PUBLISHED = "PUBLISHED"
    VERIFICATION_PENDING = "VERIFICATION_PENDING"
    VERIFIED = "VERIFIED"
    REJECTED = "REJECTED"
    FAILED = "FAILED"

    @classmethod
    def terminal_values(cls) -> tuple[str, ...]:
        """Statuses after which a retry may create a new submission row."""
        return (cls.REJECTED.value, cls.FAILED.value)


class ContentStatus(DomainEnum):
    """Review state of AI-drafted listing copy.

    Content is always persisted before submission, so a human reviews (and an
    auditor can later inspect) exactly what was sent.
    """

    DRAFT = "DRAFT"
    PENDING_REVIEW = "PENDING_REVIEW"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    SUPERSEDED = "SUPERSEDED"


# --------------------------------------------------------------------------- #
# BYOK, AI and operations
# --------------------------------------------------------------------------- #


class CredentialProviderType(DomainEnum):
    AI = "AI"
    SEARCH = "SEARCH"
    SEO_METRICS = "SEO_METRICS"
    WEBSITE_INTELLIGENCE = "WEBSITE_INTELLIGENCE"
    OTHER = "OTHER"


class CredentialStatus(DomainEnum):
    CONFIGURED = "CONFIGURED"
    VERIFIED = "VERIFIED"
    INVALID = "INVALID"
    DISABLED = "DISABLED"


class AiProvider(DomainEnum):
    """Supported LLM providers. Extending this list adds an adapter, never a
    change to business logic."""

    OPENAI = "openai"
    ANTHROPIC = "anthropic"
    GEMINI = "gemini"
    OPENROUTER = "openrouter"
    CUSTOM = "custom"


class AiPurpose(DomainEnum):
    """What a tenant's AI configuration is for.

    Business code asks for a *purpose*, never a vendor, so a tenant can use
    different providers for content generation and embeddings.
    """

    CONTENT_GENERATION = "content_generation"
    DISCOVERY = "discovery"
    QUALIFICATION = "qualification"
    EMBEDDING = "embedding"


class AiOperation(DomainEnum):
    GENERATE = "generate"
    EMBED = "embed"


class AiUsageStatus(DomainEnum):
    SUCCEEDED = "SUCCEEDED"
    FAILED = "FAILED"


class JobStatus(DomainEnum):
    PENDING = "PENDING"
    RUNNING = "RUNNING"
    SUCCEEDED = "SUCCEEDED"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"

File: frontend/src/config/enums.ts
/**
 * Runtime value lists for the backend's enums.
 *
 * The unions in `@/types/api` are compile-time only, but filter dropdowns and
 * permission matrices need iterable values. `allOf` ties the two together: if
 * the backend adds, removes or renames an enum member, `npm run codegen:api`
 * followed by `npm run typecheck` fails here rather than shipping a filter that
 * silently omits a status.
 */

import type {
  AiProvider,
  AiPurpose,
  CampaignStatus,
  ClientWebsiteStatus,
  ContentStatus,
  CredentialProviderType,
  CredentialStatus,
  LinkType,
  MembershipStatus,
  OpportunityStatus,
  OpportunityType,
  PricingType,
  PublisherCategory,
  PublisherStatus,
  SubmissionMethod,
  SubmissionStatus,
  TenantStatus,
} from "@/types/api";

/**
 * Returns the list unchanged, but only type-checks when it covers the union
 * exactly — no missing members and no members the union does not have.
 */
function allOf<U extends string>() {
  return <L extends readonly U[]>(
    list: Exclude<U, L[number]> extends never ? L : never,
  ): L => list;
}

export const CAMPAIGN_STATUSES = allOf<CampaignStatus>()([
  "DRAFT",
  "ACTIVE",
  "PAUSED",
  "COMPLETED",
  "ARCHIVED",
] as const);

export const CLIENT_WEBSITE_STATUSES = allOf<ClientWebsiteStatus>()([
  "ACTIVE",
  "PAUSED",
  "ARCHIVED",
] as const);

export const PUBLISHER_STATUSES = allOf<PublisherStatus>()([
  "DISCOVERED",
  "QUALIFYING",
  "QUALIFIED",
  "REJECTED",
  "BLOCKED",
  "ARCHIVED",
] as const);

export const PUBLISHER_CATEGORIES = allOf<PublisherCategory>()([
  "BUSINESS_DIRECTORY",
  "LOCAL_DIRECTORY",
  "COMPANY_LISTING",
  "STARTUP_DIRECTORY",
  "SOFTWARE_DIRECTORY",
  "INDUSTRY_DIRECTORY",
  "ORGANIZATION_LISTING",
  "PROFILE_LISTING",
  "REVIEW_PLATFORM",
  "EVENT_LISTING",
  "JOB_BOARD",
  "OTHER",
] as const);

export const OPPORTUNITY_STATUSES = allOf<OpportunityStatus>()([
  "DISCOVERED",
  "QUALIFYING",
  "QUALIFIED",
  "REJECTED",
  "SELECTED",
  "READY",
  "SUBMITTED",
  "PUBLISHED",
  "FAILED",
  "EXPIRED",
] as const);

export const OPPORTUNITY_TYPES = allOf<OpportunityType>()([
  "FREE_DIRECTORY_LISTING",
  "FREE_LOCAL_LISTING",
  "FREE_COMPANY_PROFILE",
  "FREE_STARTUP_LISTING",
  "FREE_SOFTWARE_LISTING",
  "FREE_INDUSTRY_LISTING",
  "FREE_ORGANIZATION_LISTING",
  "FREE_PROFILE_LISTING",
] as const);

export const SUBMISSION_STATUSES = allOf<SubmissionStatus>()([
  "READY",
  "IN_PROGRESS",
  "SUBMITTED",
  "PENDING_APPROVAL",
  "PUBLISHED",
  "VERIFICATION_PENDING",
  "VERIFIED",
  "REJECTED",
  "FAILED",
] as const);

export const SUBMISSION_METHODS = allOf<SubmissionMethod>()([
  "MANUAL",
  "FORM",
  "EMAIL",
  "API",
  "ACCOUNT_REQUIRED",
  "UNKNOWN",
] as const);

export const CONTENT_STATUSES = allOf<ContentStatus>()([
  "DRAFT",
  "PENDING_REVIEW",
  "APPROVED",
  "REJECTED",
  "SUPERSEDED",
] as const);

export const CREDENTIAL_STATUSES = allOf<CredentialStatus>()([
  "CONFIGURED",
  "VERIFIED",
  "INVALID",
  "DISABLED",
] as const);

export const CREDENTIAL_PROVIDER_TYPES = allOf<CredentialProviderType>()([
  "AI",
  "SEARCH",
  "SEO_METRICS",
  "WEBSITE_INTELLIGENCE",
  "OTHER",
] as const);

export const MEMBERSHIP_STATUSES = allOf<MembershipStatus>()([
  "ACTIVE",
  "INVITED",
  "SUSPENDED",
  "REMOVED",
] as const);

export const TENANT_STATUSES = allOf<TenantStatus>()([
  "ACTIVE",
  "SUSPENDED",
  "ARCHIVED",
] as const);

export const PRICING_TYPES = allOf<PricingType>()([
  "FREE",
  "PAID",
  "MIXED",
  "UNKNOWN",
] as const);

export const LINK_TYPES = allOf<LinkType>()([
  "DOFOLLOW",
  "NOFOLLOW",
  "MIXED",
  "UNKNOWN",
] as const);

export const AI_PROVIDERS = allOf<AiProvider>()([
  "openai",
  "anthropic",
  "gemini",
  "openrouter",
  "custom",
] as const);

export const AI_PURPOSES = allOf<AiPurpose>()([
  "content_generation",
  "discovery",
  "qualification",
  "embedding",
] as const);

/**
 * The MVP promotes free listings only (spec §73), so the publisher list and
 * every discovery request are pinned to FREE pricing. Paid, MIXED and UNKNOWN
 * exist in the backend enum but are never offered as a choice in the UI.
 */
export const MVP_PRICING_TYPE = "FREE" satisfies PricingType;

/** Opportunity statuses that mean "the human has not decided yet". */
export const OPPORTUNITY_PENDING_STATUSES: readonly OpportunityStatus[] = [
  "DISCOVERED",
  "QUALIFYING",
  "QUALIFIED",
];

/** Submission statuses that still need a reviewer. */
export const SUBMISSION_REVIEW_STATUSES: readonly SubmissionStatus[] = [
  "READY",
  "PENDING_APPROVAL",
];

/** Submission statuses past which no further action is expected. */
export const SUBMISSION_TERMINAL_STATUSES: readonly SubmissionStatus[] = [
  "VERIFIED",
  "REJECTED",
  "FAILED",
];

/* -------------------------------------------------------------------------- */
/* Statuses the backend serialises as a bare string                           */
/*                                                                            */
/* DiscoveryRunRead.status and JobRead.status are typed `str` in the OpenAPI   */
/* document rather than as enum refs, so codegen cannot narrow them. These     */
/* mirror app/core/enums.py — DiscoveryRunStatus and JobStatus.                */
/* -------------------------------------------------------------------------- */

export const DISCOVERY_RUN_STATUSES = [
  "PENDING",
  "RUNNING",
  "COMPLETED",
  "FAILED",
] as const;
export type DiscoveryRunStatus = (typeof DISCOVERY_RUN_STATUSES)[number];

export const JOB_STATUSES = [
  "PENDING",
  "RUNNING",
  "SUCCEEDED",
  "FAILED",
  "CANCELLED",
] as const;
export type JobStatus = (typeof JOB_STATUSES)[number];

/** Statuses that mean a discovery run or job is still working, so keep polling. */
export const IN_FLIGHT_STATUSES: readonly string[] = ["PENDING", "RUNNING"];

/** Statuses that mean a discovery run or job has finished, so stop polling. */
export function isTerminalRunStatus(status: string): boolean {
  return !IN_FLIGHT_STATUSES.includes(status);
}

/** Seeded role slugs (app/core/enums.py — SystemRoleSlug). */
export const SYSTEM_ROLE_SLUGS = [
  "owner",
  "admin",
  "seo_manager",
  "seo_specialist",
  "viewer",
] as const;
export type SystemRoleSlug = (typeof SYSTEM_ROLE_SLUGS)[number];
